### ISSUE TITLE
Caching on Path

### ISSUE THREAD

**Issue by RALifeCoach** [DianaIonita/serverless-api-gateway-caching#14]
Is there a way to turn on caching for the path?

**Comment by RALifeCoach** [comment]
Yes. Now how do I set up caching based on path and path parameter. For example, I have 
…/en/pool/{PoolId} and
…/en/player{PolId}.

I need to cache based on:
-name: /en/pool - request.path??? Request.path.url???
-name: request.path.poolId

and

-name: /en/player
-name: poolId

**Comment by RALifeCoach** [comment]
That isn’t working. I get the same file regardless of the final qualifier. I also tried adding the final qualifier, but still no success.

**Comment by thebeatlecoder** [comment]
My bad, I was wrong about this. If you have separate lambda functions serving each URL (which is the pattern we use), you can do this:

```yml
functions:
  get-pool:
    handler: rest_api/pool/get/handler.handle
    events:
      - http:
          path: /en/pool/{poolId}
          method: get
          caching:
            enabled: true
            cacheKeyParameters:
              - name: request.path.poold

  get-player:
    handler: rest_api/player/get/handler.handle
    events:
      - http:
          path: /en/player/{poolId}
          method: get
          caching:
            enabled: true
            cacheKeyParameters:
              - name: request.path.poold

```

I'm not sure how to do it if you have one lambda function serving all of your URLs. Is that what you have currently?

**Comment by thebeatlecoder** [comment]
If you do have a single lambda function serving all requests, maybe you can do this:

```yml
http-get-requests:
    handler: rest_api/get/handler.handle
    events:
      - http:
          path: /{shortUrl}
          method: get
          caching:
            enabled: true
            cacheKeyParameters:
              - name: request.path.shortUrl
```

I haven't tried this personally but it looks like it should work.

**Comment by thebeatlecoder** [comment]
If you do  have a single lambda function serving all our your URLs, this should work. I just tired it and it worked for me:

```yml
proxy-http-get:
    handler: rest_api/proxy/handler.handle
    events:
      - http:
          path:  /{proxy+}
          method: get
          caching:
            enabled: true
            ttlInSeconds: 3600
            cacheKeyParameters:
              - name: request.path.proxy
```

**Comment by RALifeCoach** [comment]
This is what I started with. And for both pool and player, the poolId is ignored and the first read sets the cache and that is the file returned for subsequent reads.

**Comment by RALifeCoach** [comment]
```
service: service-name

provider:
  name: aws
  runtime: nodejs8.10
  stage: ${opt:stage, 'dev'}
  iamRoleStatements:
    - Effect: 'Allow'
      Action:
        - 'lambda:InvokeFunction'
      Resource: "*"
  vpc:
    securityGroupIds:
      - ${file(./config/config.${self:provider.stage}.api.json):vpcSecurityGroup}
    subnetIds:
      - ${file(./config/config.${self:provider.stage}.api.json):vpcSubnet}
  environment:
    MODE: 'api'
    DB_HOST: ${file(./config/config.${self:provider.stage}.api.json):dbHost}
    DB_USER: ${file(./config/config.${self:provider.stage}.api.json):dbUser}
    DB_PASSWORD: ${file(./config/config.${self:provider.stage}.api.json):dbPassword}
    LOG_LEVEL: ${file(./config/config.${self:provider.stage}.api.json):logLevel}

package:
  individually: true

plugins:
  - serverless-webpack
  - serverless-offline
  - serverless-api-gateway-caching
  - serverless-plugin-warmup

custom:
  apiGatewayCaching:
    enabled: true
    clusterSize: '0.5' # defaults to '0.5'
    ttlInSeconds: 600 # defaults to the maximum allowed: 3600
    perKeyInvalidation:
      requireAuthorization: true # default is true
      handleUnauthorizedRequests: IgnoreWithWarning # default is "IgnoreWithWarning"
  webpack:
    webpackConfig: 'webpack.config.js'
    includeModules: true
    packager: 'npm'

functions:
  championship:
    handler: src/api/index.championship
    warmup: true
    events:
      - http:
          path: championship/{type}
          method: get
          cors: true
          request:
            parameters:
              paths:
                type: true
          caching:
            enabled: true
            ttlInSeconds: 300
            perKeyInvalidation:
              requireAuthorization: true
              handleUnauthorizedRequests: Ignore
            cacheKeyParameters:
              - name: request.path.type
  competition:
    handler: src/api/index.competition
    warmup: true
    events:
      - http:
          path: competition/{champId}
          method: get
          cors: true
          request:
            parameters:
              paths:
                champId: true
          caching:
            enabled: true
            ttlInSeconds: 300
            perKeyInvalidation:
              requireAuthorization: true
              handleUnauthorizedRequests: Ignore
            cacheKeyParameters:
              - name: request.path.champId
  keyCompetition:
    handler: src/api/index.keyCompetition
    warmup: true
    events:
      - http:
          path: keyCompetition/{key}
          method: get
          cors: true
          request:
            parameters:
              paths:
                key: true
          caching:
            enabled: true
            ttlInSeconds: 300
            perKeyInvalidation:
              requireAuthorization: true
              handleUnauthorizedRequests: Ignore
            cacheKeyParameters:
              - name: request.path.key
  competitionPool:
    handler: src/api/index.competitionPool
    warmup: true
    events:
      - http:
          path: competitionPool/{competitionId}
          method: get
          cors: true
          request:
            parameters:
              paths:
                competitionId: true
          caching:
            enabled: true
            ttlInSeconds: 300
            perKeyInvalidation:
              requireAuthorization: true
              handleUnauthorizedRequests: Ignore
            cacheKeyParameters:
              - name: request.path.competitionId
  pool:
    handler: src/api/index.pool
    warmup: true
    events:
      - http:
          path: pool/{poolId}
          method: get
          cors: true
          request:
            parameters:
              paths:
                poolId: true
          caching:
            enabled: true
            ttlInSeconds: 300
            perKeyInvalidation:
              requireAuthorization: true
              handleUnauthorizedRequests: Ignore
            cacheKeyParameters:
              - name: request.path.shortUrl
              - name: request.path.poolId
  player:
    handler: src/api/index.player
    warmup: true
    events:
      - http:
          path: player/{poolId}
          method: get
          cors: true
          request:
            parameters:
              paths:
                poolId: true
          caching:
            enabled: true
            ttlInSeconds: 300
            perKeyInvalidation:
              requireAuthorization: true
              handleUnauthorizedRequests: Ignore
            cacheKeyParameters:
              - name: request.path.shortUrl
              - name: request.path.poolId
```

**Comment by RALifeCoach** [comment]
Yes. They all work.

**Comment by RALifeCoach** [comment]
It is the pool and player ones the do not work.

**Comment by DianaIonita** [comment]
@RALifeCoach Have you tried the config I suggested?

**Comment by RALifeCoach** [comment]
Yes. In several different variations including exactly as you suggested. In
fact that's what I started with.

**Comment by DianaIonita** [comment]
@RALifeCoach if your code is open source and you can provide a link to it, I can try debugging. It's difficult to tell what's wrong otherwise.

**Comment by RALifeCoach** [comment]
It is not open source, but I can mock up something, if you have an AWS account where you can deploy it to.

**Comment by DianaIonita** [comment]
Yes, please provide an example where this caching plugin does not work as expected.

**Comment by RALifeCoach** [comment]
Here is a sample project: https://github.com/RALifeCoach/caching.git

**Comment by DianaIonita** [comment]
Hi @RALifeCoach,

I'm trying to reproduce the issue so I deployed your example to my AWS environment. Here are the endpoints:

```
  GET - https://0bpa2rw46m.execute-api.eu-west-1.amazonaws.com/dev/championship/{type}
  GET - https://0bpa2rw46m.execute-api.eu-west-1.amazonaws.com/dev/competition/{champId}
  GET - https://0bpa2rw46m.execute-api.eu-west-1.amazonaws.com/dev/keyCompetition/{key}
  GET - https://0bpa2rw46m.execute-api.eu-west-1.amazonaws.com/dev/competitionPool/{competitionId}
  GET - https://0bpa2rw46m.execute-api.eu-west-1.amazonaws.com/dev/pool/{poolId}
  GET - https://0bpa2rw46m.execute-api.eu-west-1.amazonaws.com/dev/player/{poolId}
```

The only thing I did, besides deploying, was to add a `response-timestamp` header in the `index.es6` file:

![image](https://user-images.githubusercontent.com/2750493/47954900-0eb3a280-df88-11e8-962e-66ac9eda7691.png)

I then used Postman to send requests and it seems like all endpoints have caching correctly configured. For instance, for the `pool/{poolId}` endpoint, if you pass in the id `some-id`, you get the same response timestamp for every request until the cache expires:

![image](https://user-images.githubusercontent.com/2750493/47954950-c8ab0e80-df88-11e8-8777-bc93ff96aead.png)

If you pass in another id, another entry in the cache store is generated so you get a new timestamp for that request.
One thing to note is that API Gateway takes a few minutes to create the cache cluster - it wouldn't be available immediately after deploying. Make sure the Cache status is set to `AVAILABLE` when you call your endpoints:

![image](https://user-images.githubusercontent.com/2750493/47954981-25a6c480-df89-11e8-8f92-579bd320b596.png)

**Comment by RALifeCoach** [comment]
"For instance, for the pool/{poolId} endpoint, if you pass in the id some-id, you get the same response timestamp for every request until the cache expires” - what if you pass in different values for poolId. Do you get different files returned?

**Comment by DianaIonita** [comment]
For example, let's say the following events happen in sequence:
1. Request to `/pool/123` returns `A`. `A` is cached for 1 hour.
2. Request to `/pool/456` returns `B`. `B` is cached for 1 hour.
3. Request to `/pool/123` returns `A` from the cache, it does not hit the deployed lambda.
4. Cache for `/pool/123` expires.
5. Request to `pool/456` returns `B` from the cache.
6. Request to `/pool/123` returns `A`. `A` is cached for 1 hour again.

Does that make sense?

**Comment by RALifeCoach** [comment]
Yes that makes sense, but is not what is happening in real life.

**Comment by DianaIonita** [comment]
Can you give an example on the endpoints I have deployed where this is not happening?

**Comment by RALifeCoach** [comment]
Sorry, I have been busy on another project.

I was able to reproduce the error using this simplified code. I opened 2 browser windows and aimed one at https://944d2c0g93.execute-api.us-east-1.amazonaws.com/qa/pool/77777 <https://944d2c0g93.execute-api.us-east-1.amazonaws.com/qa/pool/77777> and one at https://944d2c0g93.execute-api.us-east-1.amazonaws.com/qa/player/77777 <https://944d2c0g93.execute-api.us-east-1.amazonaws.com/qa/player/77777>.

I then executed both queries at the same time. After both queries finished, running https://944d2c0g93.execute-api.us-east-1.amazonaws.com/qa/pool/ <https://944d2c0g93.execute-api.us-east-1.amazonaws.com/qa/pool/>{any value} returned 77777.

**Comment by DianaIonita** [comment]
Hi @RALifeCoach,

Ok, I found the issue and released [v1.1.5](https://github.com/DianaIonita/serverless-api-gateway-caching/releases/tag/1.1.5) which fixes it. The problem happened because a function name was a substring of a different function name, so `pool` was a substring of `competitionPool`. The logic which tries to find the associated Api Gateway method from the compiled CloudFormation template and attach path parameters did not take that into account and, out of the two different methods, it just returned the first one.

Thanks for finding this and patiently walking me through the issue 👍 

All the best,
Diana

**Comment by RALifeCoach** [comment]
Thank-you for the speedy and timely work on this issue. Your efforts are much appreciated.
